stl/inc/ranges

```diff
@@ -6837,6 +6837,602 @@ namespace ranges {
+            const auto _Created_lambda = [&_Callback]<typename... _ViewTupleTypes>(_ViewTupleTypes&&... _View_tuples) {
+                (_STD invoke(_Callback, _STD forward<_ViewTupleTypes>(_View_tuples)), ...);
+            };
+
+            return _Created_lambda;
+        }
+
+        template <typename _CallbackType, typename _TupleType>
+        constexpr void _Tuple_for_each(_CallbackType&& _Callback, _TupleType&& _Tuple) noexcept(
+            noexcept(_STD apply(_Zip::_Create_tuple_for_each_lambda(_STD forward<_CallbackType>(_Callback)),
+                _STD forward<_TupleType>(_Tuple)))) /* strengthened */ {
+            _STD apply(_Zip::_Create_tuple_for_each_lambda(_STD forward<_CallbackType>(_Callback)),
+                _STD forward<_TupleType>(_Tuple));
+        }
+
+        template <size_t _DesiredOffset, size_t _FirstIndex, size_t... _OtherIndices>
+        struct _Offset_index_sequence_solver {
+            static constexpr auto _Solved_index_sequence{
+                _Offset_index_sequence_solver<_DesiredOffset, _OtherIndices...>::_Solved_index_sequence};
+        };
+
+        template <size_t _DesiredOffset, size_t... _OtherIndices>
+        struct _Offset_index_sequence_solver<_DesiredOffset, _DesiredOffset, _OtherIndices...> {
+            static constexpr auto _Solved_index_sequence{index_sequence<_DesiredOffset, _OtherIndices...>{}};
+        };
+
+        template <size_t _OriginalSequenceSize, size_t _Offset>
+            requires(_Offset <= _OriginalSequenceSize)
+        _CONSTEVAL auto _Get_offset_index_sequence() noexcept {
+            constexpr auto _Create_index_sequence_lambda = []<size_t... _Indices>(index_sequence<_Indices...>) {
+                return _Offset_index_sequence_solver<_Offset, _Indices...>::_Solved_index_sequence;
+            };
+
+            return _Create_index_sequence_lambda(make_index_sequence<_OriginalSequenceSize>{});
+        }
+
+        template <bool _IsConst, typename... _Views>
+        concept _All_random_access = (random_access_range<_Maybe_const<_IsConst, _Views>> && ...);
+
+        template <bool _IsConst, typename... _Views>
+        concept _All_bidirectional = (bidirectional_range<_Maybe_const<_IsConst, _Views>> && ...);
+
+        template <bool _IsConst, typename... _Views>
+        concept _All_forward = (forward_range<_Maybe_const<_IsConst, _Views>> && ...);
+
+        template <typename... _ViewTypes>
+        concept _Allow_views = (input_range<_ViewTypes> && ...) && (view<_ViewTypes> && ...)
+                            && (sizeof...(_ViewTypes) > 0);
+    } // namespace _Zip
+
+    template <input_range... _ViewTypes>
+        requires _Zip::_Allow_views<_ViewTypes...>
+    class zip_view : public view_interface<zip_view<_ViewTypes...>> {
+    public:
+        template <bool>
+        class _Iterator;
+
+        template <bool>
+        class _Sentinel;
+
+    private:
+        template <bool>
+        friend class _Sentinel;

```

Comment:
I believe that this is unnecessary - C++ has a rule that "nested classes are members and thus have the same access rights as any other member", i.e. a nested class can access everything within the parent class.

Reply:
I'll be the first to admit that I didn't know this before you pointed it out. That's actually quite helpful!